From f4abdb0c1d24751523e455a8a6426c26d9e7b13d Mon Sep 17 00:00:00 2001
From: Dimitar Dimitrov <dimitar@dinux.eu>
Date: Thu, 19 Jul 2018 22:08:51 +0300
Subject: [PATCH 7/7] PRU: Fix HW register numbering for DWARF2 output

A small rework of the PRU GCC port exposed that CIE data alignment is
erroneously set to 4 for PRU. In fact PRU stack is aligned to 1.
Set it to -1, to allow output from GCC to be assembled without errors.

Also, while at it, set DWARF2 HW register numbering to follow latest GCC port.

gas/ChangeLog:

2018-07-27  Dimitar Dimitrov  <dimitar@dinux.eu>

	* config/tc-pru.c (pru_regname_to_dw2regnum): Return the starting HW
	byte-register number.
	(pru_frame_initial_instructions): Use byte-numbering for FP index.
	* config/tc-pru.h (DWARF2_DEFAULT_RETURN_COLUMN): Use number from
	latest GCC.
	(DWARF2_CIE_DATA_ALIGNMENT): Set to -1.

Signed-off-by: Dimitar Dimitrov <dimitar@dinux.eu>
---
 gas/config/tc-pru.c | 17 ++++++++++++++---
 gas/config/tc-pru.h | 10 ++++++++--
 2 files changed, 22 insertions(+), 5 deletions(-)

diff --git a/gas/config/tc-pru.c b/gas/config/tc-pru.c
index 5121b1f8a4..d32f793548 100644
--- a/gas/config/tc-pru.c
+++ b/gas/config/tc-pru.c
@@ -1920,14 +1920,25 @@ pru_cons_fix_new (fragS *frag, int where, unsigned int nbytes,
 }
 
 /* Implement tc_regname_to_dw2regnum, to convert REGNAME to a DWARF-2
-   register number.  */
+   register number.  Return the starting HW byte-register number.  */
 int
 pru_regname_to_dw2regnum (char *regname)
 {
+  static const unsigned int regstart[RSEL_NUM_ITEMS] = {
+	  [RSEL_7_0]	= 0,
+	  [RSEL_15_8]	= 1,
+	  [RSEL_23_16]	= 2,
+	  [RSEL_31_24]	= 3,
+	  [RSEL_15_0]	= 0,
+	  [RSEL_23_8]	= 1,
+	  [RSEL_31_16]	= 2,
+	  [RSEL_31_0]	= 0,
+  };
+
   struct pru_reg *r = pru_reg_lookup (regname);
   if (r == NULL)
     return -1;
-  return r->index;
+  return r->index * 4 + regstart[r->regsel];
 }
 
 /* Implement tc_cfi_frame_initial_instructions, to initialize the DWARF-2
@@ -1935,7 +1946,7 @@ pru_regname_to_dw2regnum (char *regname)
 void
 pru_frame_initial_instructions (void)
 {
-  const unsigned fp_regno = 4;
+  const unsigned fp_regno = 4 * 4;
   cfi_add_CFA_def_cfa (fp_regno, 0);
 }
 
diff --git a/gas/config/tc-pru.h b/gas/config/tc-pru.h
index bc4785154b..c0239b5f9a 100644
--- a/gas/config/tc-pru.h
+++ b/gas/config/tc-pru.h
@@ -138,8 +138,14 @@ extern int pru_validate_fix_sub (struct fix *);
 
 /* We want .cfi_* pseudo-ops for generating unwind info.  */
 #define TARGET_USE_CFIPOP 1
-#define DWARF2_DEFAULT_RETURN_COLUMN 31
-#define DWARF2_CIE_DATA_ALIGNMENT (-4)
+
+/* Program Counter register number is not defined by TI documents.  Pick the
+   virtual number used by GCC.  */
+#define DWARF2_DEFAULT_RETURN_COLUMN 132
+
+/* The stack grows down, and is only byte aligned.  */
+#define DWARF2_CIE_DATA_ALIGNMENT -1
+
 #define tc_regname_to_dw2regnum pru_regname_to_dw2regnum
 extern int pru_regname_to_dw2regnum (char *regname);
 #define tc_cfi_frame_initial_instructions  pru_frame_initial_instructions
-- 
2.11.0

